src/ty_check/mod.rs

```diff
@@ -2258,36 +3394,163 @@ impl TyChecker {
+                self.ty_well_formed(&kind_ctx, ty_ctx, Some(*exec), ret_ty)?;
+                let mut visitor = FreeKindedIdents::new();
+                for param_ty in param_tys {
+                    //Add free implicit Lifetime-parameter-idents to kind_ctx
+                    visitor.set.clear();
+                    visitor.bound_idents.clear();
+                    visitor.visit_ty(param_ty);
+                    let kind_ctx = kind_ctx.clone().append_idents(
+                        visitor
+                            .set
+                            .iter()
+                            .filter(|ident| {
+                                ident.kind == Kind::Provenance && ident.ident.is_implicit
+                            })
+                            .map(|ident| ident.clone())
+                            .collect(),
+                    );
+
+                    self.ty_well_formed(&kind_ctx, ty_ctx, Some(*exec), param_ty)?;
+                }
+            }
+            TyKind::Dead(_) => {}
+        }
+
+        Ok(())
+    }
+
+    // TODO respect memory
+    fn dty_well_formed(
+        &self,
+        kind_ctx: &KindCtx,
+        ty_ctx: &TyCtx,
+        exec: Option<Exec>,
+        dty: &DataTy,
+    ) -> TyResult<()> {
+        match &dty.dty {
+            // TODO variables of Dead types can be reassigned. So why do we not have to check
+            //  well-formedness of the type in Dead(ty)? (According paper).
+            DataTyKind::Scalar(_)
+            | DataTyKind::Atomic(_)
+            | DataTyKind::Range
+            | DataTyKind::RawPtr(_)
+            | DataTyKind::Dead(_) => {}
+            DataTyKind::Ident(ident) => {
+                if !ident.is_implicit && !kind_ctx.ident_of_kind_exists(ident, Kind::DataTy) {
+                    Err(CtxError::KindedIdentNotFound(ident.clone()))?
+                }
+            }
+            DataTyKind::Struct(struct_mono_ty) => {
+                let StructDataType {
+                    name,
+                    attributes,
+                    generic_args,
+                } = struct_mono_ty;
+                let struct_decl = self.gl_ctx.struct_by_name(name)?;
+                Self::check_args_have_correct_kinds(&struct_decl.generic_params, generic_args)?;
+                iter_TyResult_to_TyResult!(attributes
+                    .iter()
+                    .map(|field| self.dty_well_formed(kind_ctx, ty_ctx, exec, &field.ty)))?;
+            }
+            DataTyKind::Ref(Provenance::Value(prv), own, mem, dty) => {
+                self.memory_well_formed(kind_ctx, mem)?;
+                let elem_ty = dty.as_ref();
+                if !elem_ty.is_fully_alive() {
+                    return Err(TyError::ReferenceToDeadTy);
+                }
+                let loans = ty_ctx.loans_in_prv(prv)?;
+                if !loans.is_empty() {
+                    let mut exists = false;
+                    //TODO what to do if exec is None?
```

Comment:
Was ist mit der Well-formedness für Referenzen, die nicht in einem exec auftreten?
Kann man dann die exec-spezifische Teil der Well-formedness einfach ignorieren